opte/src/engine/icmpv6.rs

```diff
@@ -172,3 +176,296 @@ impl HairpinAction for Icmpv6EchoReply {
+        //
+        // TODO-completeness: Check IP Hop Limit and ICMP length / option length
+
+        let flags = if self.managed_cfg {
+            NdiscRouterFlags::MANAGED
+        } else {
+            NdiscRouterFlags::empty()
+        };
+        let advert = NdiscRepr::RouterAdvert {
+            hop_limit: u8::MAX,
+            flags,
+            // Use the maximum advertised lifetime as a default router.
+            router_lifetime: Duration::from_secs(9_000),
+            // Do not specify the reachable or retrans time. Guests will decide
+            // that for themselves at this point.
+            reachable_time: Duration::from_millis(0),
+            retrans_time: Duration::from_millis(0),
+            lladdr: Some(RawHardwareAddress::from_bytes(&self.mac.bytes())),
+            mtu: Some(1500),
+            // Indicate that there are no addresses considered on-link, other
+            // than the router's advertised link-local address. This will
+            // require all traffic from the guest to go through OPTE.
+            prefix_info: None,
+        };
+        let reply = Icmpv6Repr::Ndisc(advert);
+
+        let reply_len = reply.buffer_len();
+        let mut ulp_body = vec![0u8; reply_len];
+        let mut advert_reply = Icmpv6Packet::new_unchecked(&mut ulp_body);
+        let mut csum = Csum::ignored();
+        csum.icmpv6 = Checksum::Tx;
+        reply.emit(
+            &IpAddress::Ipv6((*self.ip()).into()),
+            &src_ip,
+            &mut advert_reply,
+            &mut csum,
+        );
+
+        let mut ip = Ipv6Hdr::from(&Ipv6Meta {
+            src: *self.ip(),
+            // Safety: We match on this being Some(_) above, so unwrap is safe.
+            dst: meta.inner_ip6().unwrap().src,
+            proto: Protocol::ICMPv6,
+        });
+
+        // There are no extension headers, so the ULP is the only content.
+        ip.set_pay_len(reply_len as u16);
+
+        // The Ethernet frame should come from OPTE's virtual gateway MAC, and
+        // be destined for the guest which sent us the packet.
+        let eth = EtherHdr::from(&EtherMeta {
+            dst: self.src_mac.into(),
+            src: self.mac.into(),
+            ether_type: ether::ETHER_TYPE_IPV6,
+        });
+
+        let mut pkt_bytes =
+            Vec::with_capacity(ETHER_HDR_SZ + IPV6_HDR_SZ + reply_len);
+        pkt_bytes.extend_from_slice(&eth.as_bytes());
+        pkt_bytes.extend_from_slice(&ip.as_bytes());
+        pkt_bytes.extend_from_slice(&ulp_body);
+        Ok(AllowOrDeny::Allow(Packet::copy(&pkt_bytes)))
+    }
+}
+
+impl HairpinAction for NeighborAdvertisement {
+    fn implicit_preds(&self) -> (Vec<Predicate>, Vec<DataPredicate>) {
+        let hdr_preds = vec![
+            // We expect that the source MAC is the MAC provided to the Guest.
+            Predicate::InnerEtherSrc(vec![EtherAddrMatch::Exact(self.src_mac)]),
+            // It's directed to the virtual gateway's MAC address.
```

Comment:
I thought NS messages were directed to the all-nodes multicast address?

Reply:
Okay, so reading more I see there is unicast and multicast solicitation. So this kind of comes back to my various comments about the general vs VPC-specific wording in the comments. Now I think we might also be starting to creep into VPC-specific implementation. That said, I don't want to cause a stir in this PR and stop momentum. I just want to share that my intention (which is probably not clear outside of my own mind) is that opte-api holds actions that are mostly generic, and anything that starts getting fairly specific to oxide-vpc lives in there. I realize that you are just following the pattern I've laid out and that this action is borderline general vs specific. You could also argue that we only have the one implementation at the moment, so maybe this is all fine for now. I just wanted to point out this division that I try to keep in my head and that I've been trying to enforce more and more in the code.

Reply:
It's a good question about this particular case. I had assumed that an NS would only be received _after_ an RS, in which case the client would have a MAC address for the virtual gateway in their cache. At that point, clients are allowed to unicast the following NS or multicast it, as you pointed out.

The other reason I ignored the multicast is that we will drop any NS for _another_ neighbor. That is, we only want to respond to NSs with a target node of the gateway, and everything else needs to be determined as _off-link_. So the gateway could respond to solicited-node multicast NSs, assuming that the target node in that message is the gateway IPv6 address.

That could be an optimization though. I mean, clients could still be multicasting NSs, and the gateway _should_ respond if the target node is its IPv6 address.

This actually brings up a second point, something I think I missed. During Duplicate Address Detection, the client will be sending out NSs that _are_ multicast to the solicited-node multicast group defined by their candidate link-local IPv6 address. OPTE should mostly ignore those, with one exception: On the off-chance a client tries to claim _our_ IPv6 address, we need to wave them off!

All this is to say, I think we should probably add support for responding with NAs when the client sends a multicast NS that has our IPv6 address as the target node. This will force the client to pick a new link-local there's a duplicate, and is probably the expected behavior, even _if_ the client has our MAC in their cache.

Reply:
> This actually brings up a second point, something I think I missed. During Duplicate Address Detection, the client will be sending out NSs that are multicast to the solicited-node multicast group defined by their candidate link-local IPv6 address. OPTE should mostly ignore those, with one exception: On the off-chance a client tries to claim our IPv6 address, we need to wave them off!

So I'm not to worried about DAD here, as I would expect all guest OSs we care about use the EUI-64 of the NIC's MAC address to compute the link-local, right? However, reading 7.2.2, I get the feeling that a node is going to use the solicited-node multicast address to send an NS to the gateway (opte).

```
[7.2.2](https://www.rfc-editor.org/rfc/rfc2461#section-7.2.2).  Sending Neighbor Solicitations

   When a node has a unicast packet to send to a neighbor, but does not
   know the neighbor's link-layer address, it performs address
   resolution.  For multicast-capable interfaces this entails creating a
   Neighbor Cache entry in the INCOMPLETE state and transmitting a
   Neighbor Solicitation message targeted at the neighbor.  The
   solicitation is sent to the solicited-node multicast address
   corresponding to the target address.
```

Granted, I suppose our guest could choose to not have a multicast-capable interface, but I'm assuming that's always on by default (I could be wrong about that). Though, if you tested this, it must have worked, right? So I guess the guest did use unicast to send the NS to the gateway.

To me it seems like the safest bet is to look for a NS on both the unicast and solicited-node multicast address. That way we reply to the guest no matter how it decides to perform NS for its gateway.

Reply:
> So I'm not to worried about DAD here, as I would expect all guest OSs we care about use the EUI-64 of the NIC's MAC address to compute the link-local, right?

My understanding that while this is the most common, there are some flavors of Linux (RHEL and Fedora) that may be configured to use a "privacy-preserving" link-local address. I really don't understand how that preserves any privacy, since the Ethernet frames will by definition have the MAC address in them, so you don't lose any more privacy by using that in an EUI-64 transform. But I gather it could happen.

> To me it seems like the safest bet is to look for a NS on both the unicast and solicited-node multicast address.

Agreed. I'll add the solicited-node multicast group to the predicates of OPTE's NA. Guests should get back the right NA no matter how they try to resolve it (unicast or multicast). Thanks for walking through that.

Reply:
This ended up being a bit bigger than I expected. I beefed up the validation of NS messages from the client, and the generation of the NA from us on the basis of that NS data. I really just missed this in the initial implementation, and did not read the relevant RFC 2841 sections closely enough.

I pulled the NS validation into [this function](https://github.com/oxidecomputer/opte/blob/ndp-support/opte/src/engine/icmpv6.rs#L359), which is based off [section 7.1.1](https://www.rfc-editor.org/rfc/rfc4861.html#section-7.1.1).

I pulled the NA generation into [this function](https://github.com/oxidecomputer/opte/blob/ndp-support/opte/src/engine/icmpv6.rs#L424), which is based of [section 7.2.4](https://www.rfc-editor.org/rfc/rfc4861.html#section-7.2.4).